### PR TITLE
mgr/dashboard: add backend-driven UI tables

### PR DESCRIPTION
**Pull request by epuertat**
As an example this PR displays the list of Ceph Auth users, keys and caps. The code required for that is [just this](https://github.com/ceph/ceph/pull/43771/files#diff-ff451c1362527b76017fba7d7e74136abcacc06a7a0ca1326d784789e798d820R13-R34).

image

The goal of this PR is to explore the feasibility of backend-driven UI development with the least possible amount of UI code involved.

It tries to minimize the amount of UI code required by feeding a generic table-like component (RestTable) with backend-generated JSON data.

Additionally it adds a 'data gateway' service that acts as an endpoint broker between the UI and the REST API (e.g.: `api.cluster.user@1.0`). This is needed in order to avoid the need of having per-endpoint Angular services, which often include duplicaton, redundant logic/tests, etc. The metadata endpoint is available at (`ui-api.cluster.user@1.0`) but there should be other/better ways to make resource metadata available (embedded in the main JSON, in HTTP headers, etc).

This is just a proof of concept and there's lot of room for improvement.

Fixes: https://tracker.ceph.com/issues/52701
Signed-off-by: Ernesto Puerta <epuertat@redhat.com>

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug
- [x] Generates a UI table page from a back-end resource and a generic Table component, including table metadata.
- [x] Back-end endpoints are now model-centric, instead of the previous procedural (model-less) approach.
- [ ] Discover and [dynamically generate Angular routes](https://rmannibucau.metawerx.net/angular-ng-10-create-update-route-at-runtime.html) from a back-end endpoint.
- [ ] Feed Component Input properties [from Router params](https://medium.com/@andrewcherepovskiy/passing-route-params-into-angular-components-input-properties-fc85c34c9aca)
- [ ] 

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sherrif10** [comment]
Hello @epuertat . Are you still working on this.

**Comment by epuertat** [comment]
> Hello @epuertat . Are you still working on this.

Hey @sherrif10, nope, that's abandoned for the time being. If you're interested in taking over, all yours. And if you need any help/context I'll gladly give it to you. Thanks!

**Comment by sherrif10** [comment]
Thanks @epuertat

**Comment by pereman2** [comment]
as per last commit:
image

**Comment by nizamial09** [comment]
jenkins test dashboard

**Comment by pereman2** [comment]
jenkins test make check

**Review by nizamial09 — APPROVED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/shared/models/crud-table-metadata.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { TemplateRef } from '@angular/core';
+class TableColumn {
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/models/crud-table-metadata.ts:2]: we have some models there already like https://github.com/ceph/ceph/blob/main/src/pybind/mgr/dashboard/frontend/src/app/shared/models/cd-table-column.ts. can we reuse them or extend them here? @pereman2

**Comment by nizamial09** [comment]
tests are happy
```
  Cluster Users
    breadcrumb and tab tests
      ✓ should open and show breadcrumb
    Cluster users table
      ✓ should verify the table is not empty
      ✓ should verify the keys are hidden
```

File: src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/crud-table/crud-table.component.html [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  ></cd-table>
+</ng-container>
+
+<ng-template #badgeDictTpl
+             let-value="value">
+  <span *ngFor="let instance of value | keyvalue; last as isLast">
+    <span class="badge badge-background-primary" >{{ instance.key }}: {{ instance.value }}</span>
+    <ng-container *ngIf="!isLast">&nbsp;</ng-container>
+  </span>
+</ng-template>
```
Comment (epuertat) [inline src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/crud-table/crud-table.component.html:16]: I think we already have (at least) 2 different templates to display badges.

**Review by epuertat — COMMENTED** [review]
Great job @pereman2 ! There are still some things over there that could/should be polished in order to make this fully reusable. Let me know your thoughts!

File: src/pybind/mgr/dashboard/controllers/_crud.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        return {k: serialize(v, hints[k]) for k, v in zip(o._fields, o)}
+    elif isinstance(o, (list, tuple, set)):
+        # json serializes list and tuples to arrays, hence we also serialize
+        # sets to lists.
+        # NOTE: we could add a metadata value in a list to indentify tuples and,
+        # sets if we wanted but for now let's go for lists.
+        return [serialize(i) for i in o]
+    elif isinstance(o, (Iterator, Generator)):
+        return [serialize(i) for i in o]
+    elif expected_type and issubclass(expected_type, SecretStr):
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/pybind/mgr/dashboard/controllers/_crud.py:31]: The naming might not be the best here but it intends to add info about with what type the variable was set. `SecretStr` is also an `str` so you want to treat `SecretStr` and `str` in different ways.

[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/_crud.py:31]: [Ref](https://docs.python.org/3/library/typing.html#typing.NamedTuple):
```py
>>> User('foo').__annotations__['s']
<class '__main__.SecretStr'>
```

[the remark this answers is not in the document]
Comment (pereman2) [reply src/pybind/mgr/dashboard/controllers/_crud.py:31]: > [Ref](https://docs.python.org/3/library/typing.html#typing.get_type_hints):
> 
> ```python
> >>> typing.get_type_hints(User('foo'))['secret']
> <class '__main__.SecretStr'>
> ```

I know that's why I used `expected_type` and `get_type_hints` in serializing NamedTuples

**Comment by zdover23** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/controllers/_crud.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (epuertat) [reply src/pybind/mgr/dashboard/controllers/_crud.py:31]: Yeah, part of the issue is that the Python static typing model is confusing. For this to work, when instantiating the tuple, it should actually initialize the named tuple with the proper input args:
```py
class SecretStr(str): pass

class User(NamedTuple):
  s: SecretStr

u = User(SecretStr('foo'))
>>> type(u.s)
<class '__main__.SecretStr'>
```

As we are already defining the types, we could leverage the info in the `_annotations_` attribute to initialize the input args with the expected types (dynamic casting). However, the `NamedTuple` doesn't support extending the `__init__` method and has [an extremely obscure implementation](https://github.com/python/cpython/blob/89f52293281b6efc4ef666ef25e677683821f4b9/Lib/typing.py#L2827-L2901).

A possibility would be [to use a metaclass to extend it](https://stackoverflow.com/questions/50367661/customizing-typing-namedtuple). What do you think about this?

**Comment by pereman2** [comment]
jenkins test make check

**Comment by pereman2** [comment]
jenkins test dashboard

**Comment by pereman2** [comment]
jenkins test dashboard cephadm

**Comment by pereman2** [comment]
jenkins test dashboard

**Comment by nizamial09** [comment]
jenkins test dashboard

**Review by pereman2 — APPROVED** [review]
gj

**Comment by ljflores** [comment]
Hey @epuertat, @nizamial09, could this PR be in any way related to https://tracker.ceph.com/issues/57345?

At first, I suspected https://github.com/ceph/ceph/commit/05c1b776c29795af6d015dd14aae2aa1f5d1b912#diff-bd84516dac4dc4b76ed366cf95b2d1c3dedc23c3c03a5f58dd0aa0b03e31eec5, but that was merged only 13 hours ago, and the merge of this PR coincides with when the Tracker was raised.

Might not be related though-- I'm still investigating the issue.

**Comment by epuertat** [comment]
@ljflores @pereman2 is investigating this and according to him it's related to Python 3.10 (base interpreter in Ubuntu Jammy) breaking changes.

**Comment by pereman2** [comment]
@epuertat @ljflores https://github.com/ceph/ceph/pull/48032

**Pull request merged**